src/algebra/square.lean

```diff
@@ -0,0 +1,201 @@
+
+lemma square_iff_factorization_even {m : ℕ} :
+square m ↔ ∀ (p : ℕ), even (m.factorization p) :=
+begin
+  by_cases hm_zero : m = 0,
+  { simp only [hm_zero, square_zero, factorization_zero, finsupp.coe_zero,
+      pi.zero_apply, forall_const, true_iff],
+    use 0,
+    exact mul_zero 2, },
+  { split,
+    { rintros ⟨c, hc⟩,
+      by_cases hc' : c = 0,
+      { rw [hc', mul_zero] at hc,
+        rw ←hc,
+        simp only [factorization_zero, finsupp.coe_zero, pi.zero_apply, forall_const],
+        use 0,
+        exact mul_zero 2, },
+      { intros p,
+        rw [←hc, nat.factorization_mul hc' hc', finsupp.coe_add, ←two_mul],
+        simp [even], }, },
+    { intros hp,
+      use m.factorization.prod (λ p i, p ^ (i / 2)),
+      rw ← finsupp.prod_mul,
+      conv { to_rhs, rw ← nat.factorization_prod_pow_eq_self hm_zero,},
+      apply finsupp.prod_congr,
+      intros x hx,
+      rcases hp x with ⟨c, hc⟩,
+      rw [←pow_add, ←two_mul, hc],
+      congr,
+      simp, }, },
+end
+
+/-- The property of being a square is multiplicative. The ← direction can be generalized
+to an arbitrary commutative monoid. See `square_mul_of_square_of_square`. Similarly,
+this lemma could be generalized to an arbitrary unique factorization domain, but we make use
+of `nat.factorization` in this proof. -/
+lemma square_mul {m n : ℕ} (hmn : m.coprime n) :
+  square (m * n) ↔ square m ∧ square n :=
+begin
+  split,
+  { rw [square_iff_factorization_even, square_iff_factorization_even],
+    rw square_iff_factorization_even,
+    intros hmn',
+    split,
+    { intros p,
+      by_cases hp : p ∈ m.factors,
+      { rw ← factorization_eq_of_coprime_left hmn hp, exact hmn' p, },
+      { rw [←nat.factors_count_eq, list.count_eq_zero_of_not_mem hp],
+        use zero,
+        exact mul_zero 2, }, },
+    { intros p,
+      by_cases hp : p ∈ n.factors,
+      { rw ← factorization_eq_of_coprime_right hmn hp, exact hmn' p, },
+      { rw [←nat.factors_count_eq, list.count_eq_zero_of_not_mem hp],
+        use zero,
+        exact mul_zero 2, }, }, },
+  { rintros ⟨hm, hn⟩, exact square_mul_of_square_of_square hm hn, },
+end
+
+lemma square_prime_pow_iff_pow_even : ∀ (p i : ℕ), nat.prime p → (square (p ^ i) ↔ even i) :=
+begin
+  intros p i hp,
+  split,
+  { rintros ⟨s, hs⟩,
+    have : s ∣ p ^ i,
+      calc s ∣ s * s : dvd_mul_left s s
+        ... = p ^ i : hs,
+    rw nat.dvd_prime_pow hp at this,
+    rcases this with ⟨k, hk, s_eq⟩,
+    rw [s_eq, ←pow_add] at hs,
+    have aa : (p ^ (k + k)).factorization p = k + k, { simp [hp.factorization_pow], },
+    have bb : (p ^ i).factorization p = i, { simp [hp.factorization_pow], },
+    have : i = k + k, { rw [←aa, ←bb, hs], },
+    rw this,
+    use k,
+    exact (two_mul k).symm, },
+  { rintros ⟨k, hk⟩,
+    use p ^ k,
+    rw [hk, pow_mul, pow_two, mul_pow], },
+end
+
+/-- If `x` is a square, return its square root. Else, return 0. -/
+def sqrt0 (x : ℕ) := ite (square x) (sqrt x) 0
```

Comment:
Lemmas `sqrt0_square` and `sqrt0_not_square` would come in handy here. You could use them to golf pretty much every single proof coming after.

Reply:
Also, `sqrt0 x = 0 ↔ x = 0 ∨ ¬ square x` might come in handy sometime.

Reply:
By `sqrt0_square` you mean to break the `ite` up into two lemmas like the following?

```lean
lemma sqrt0_square {s : ℕ} (hsquare : square s) : sqrt0 s = sqrt s
```

Also, should that be called `sqrt0_of_square`? Still not totally comfortable with the naming conventions!

Reply:
Yep, that's what I meant. `of` is reserved for implications, while this lemma is more of an equality, so I'd rather not use it.

Reply:
Awesome! Done.